src/jrd/tpc.cpp

```diff
@@ -130,11 +130,11 @@ void TipCache::finalizeTpc(thread_db* tdbb)
 	// To avoid race conditions, this function might only
 	// be called during database shutdown when AST delivery is already disabled
 
-	// wait for all initializing processes (PR)
+	// wait for other processes
 	Lock lock(tdbb, 0, LCK_tpc_init);
 
-	if (!LCK_lock(tdbb, &lock, LCK_SW, LCK_WAIT))
-		ERR_bugcheck_msg("Unable to obtain TPC lock (SW)");
+	if (!LCK_lock(tdbb, &lock, LCK_EX, LCK_WAIT))
+		ERR_bugcheck_msg("Unable to obtain TPC lock (EX)");
 
```

Comment:
Are we need serialization here ? Any better ways ?
Same in initializeTpc() few lines below

Reply:
Probably it can be replaced with atomic access to mhb_flags. But before doing that - are we sure such serialization is a problem?

Reply:
I think - yes, it will be a problem when attach\dettach calls rate is high. 
Probably we need some perf testing.